### PR TITLE
4 new English providers.

### PR DESCRIPTION
**Pull request by KillerDogeEmpire**
I worked with Adippe and he helped me make these providers. They work perfectly. The reason there is so many tags in the Readfrom.net provider is because it's not possible to get the synopsis and tags in the book. Due to how the website works. Other than that downside these are solid providers. Readfrom.net has more books than readanybook. They all have home pages too

Screenshot_2022-05-01-20-55-09

### CODE REVIEW

**Comment by KillerDogeEmpire** [comment]
Crap it looks like freewebnovel just died and all the novels for some reason have 0 chapters

**Comment by KillerDogeEmpire** [comment]
> Crap it looks like freewebnovel just died and all the novels for some reason have 0 chapters

I'm using the same code that was working earlier but now all the novels have  0 chapters

**Comment by AhmadNaruto** [comment]
> > Crap it looks like freewebnovel just died and all the novels for some reason have 0 chapters
> 
> I'm using the same code that was working earlier but now all the novels have 0 chapters

I also haven't found the solution but the ktp doesn't render javascript . that's why all chapters are 0.

**Comment by AhmadNaruto** [comment]
> > Crap it looks like freewebnovel just died and all the novels for some reason have 0 chapters
> 
> I'm using the same code that was working earlier but now all the novels have 0 chapters

iam done with fix tag when select with jsoup.

File: app/src/main/java/com/lagradost/quicknovel/providers/AllNovelProvider.kt [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        val firstdocument = Jsoup.parse(firstresponse.text)
+        fun get_id (tagvalue:String?): String? {
+            for (i in firstdocument.select("#hot-genre-select>option")){
+                if (i.text()==tagvalue){
+                    return i.attr("value")
+                }
+            }
+            return null
+        }
+        val url = "https://allnovel.org/ajax-search?type=hot&genre=${get_id(tag)}"
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/quicknovel/providers/AllNovelProvider.kt]: use $mainUrl instead

File: app/src/main/java/com/lagradost/quicknovel/providers/AllNovelProvider.kt [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    override fun loadHtml(url: String): String? {
+        val response = khttp.get(url)
+        val document = Jsoup.parse(response.text)
+        return document.selectFirst("#chapter-content").html().replace(" If you find any errors ( broken links, non-standard content, etc.. ), Please let us know &lt; report chapter &gt; so we can fix it as soon as possible."," ")
+    }
+
+
+    override fun search(query: String): List<SearchResponse> {
+        val response =
+            khttp.get("https://allnovel.org/search?keyword=$query") // AJAX, MIGHT ADD QUICK SEARCH
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/quicknovel/providers/AllNovelProvider.kt]: use $mainUrl

File: app/src/main/java/com/lagradost/quicknovel/providers/AzynovelProvider.kt [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    override fun loadHtml(url: String): String? {
+        val response = khttp.get(url)
+        val document = Jsoup.parse(response.text)
+        return document.selectFirst("div.column.is-9 > div:nth-child(5)").html()
+    }
+
+
+    override fun search(query: String): List<SearchResponse> {
+        val response =
+            khttp.get("https://azynovel.com/search?q=$query") // AJAX, MIGHT ADD QUICK SEARCH
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/quicknovel/providers/AzynovelProvider.kt]: $mainUrl

File: app/src/main/java/com/lagradost/quicknovel/providers/FreeWebNovelProvider.kt [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    override fun loadHtml(url: String): String? {
+        val response = khttp.get(url)
+        val document = Jsoup.parse(response.text)
+        return document.selectFirst("div.txt").html()
+    }
+
+
+    override fun search(query: String): List<SearchResponse> {
+        val response = khttp.post(
+            mainUrl+"search/",
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/quicknovel/providers/FreeWebNovelProvider.kt]: $mainUrl not +

File: app/src/main/java/com/lagradost/quicknovel/providers/FreeWebNovelProvider.kt [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+import com.lagradost.quicknovel.DataStore.toKotlinObject
+import org.jsoup.Jsoup
+import org.jsoup.select.Elements
+import java.lang.Exception
+import java.util.*
+import kotlin.collections.ArrayList
+
+class FreeWebNovelProvider : MainAPI() {
+    override val name = "FreeWebNovel"
+    override val mainUrl = "https://freewebnovel.com/"
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/quicknovel/providers/FreeWebNovelProvider.kt]: mainUrl should not end with /

File: app/src/main/java/com/lagradost/quicknovel/providers/AzynovelProvider.kt [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+import com.lagradost.quicknovel.DataStore.toKotlinObject
+import org.jsoup.Jsoup
+import org.jsoup.select.Elements
+import java.lang.Exception
+import java.util.*
+import kotlin.collections.ArrayList
+
+class AzynovelProvider : MainAPI() {
+    override val name = "AzyNovel"
+    override val mainUrl = "https://azynovel.com/"
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/quicknovel/providers/AzynovelProvider.kt]: mainUrl should not end with /

File: app/src/main/java/com/lagradost/quicknovel/providers/FreeWebNovelProvider.kt [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        }
+        val synopsis = document.selectFirst("div.inner").text()
+
+        val data: ArrayList<ChapterData> = ArrayList()
+        val chapternumber0 = document.select("div.m-newest1 > ul.ul-list5 > li")[1]
+        val chapternumber1 = chapternumber0.selectFirst("a").attr("href")
+        val aid = "[0-9]+s.jpg".toRegex().find(response.text)?.value?.substringBefore("s")
+        val acode = "(?<=r_url\" content=\"https://freewebnovel.com/)(.*)(?=/chapter)".toRegex().find(response.text)?.value
+        val chaptersDataphp = khttp.post(
+            "https://freewebnovel.com/api/chapterlist.php",
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/quicknovel/providers/FreeWebNovelProvider.kt]: $mainUrl

File: app/src/main/java/com/lagradost/quicknovel/providers/ReadfromnetProvider.kt [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+    override fun loadHtml(url: String): String? {
+        val response = khttp.get(url)
+        val document = Jsoup.parse(response.text)
+        return document.selectFirst("#textToRead").html()
+    }
+
+
+    override fun search(query: String): List<SearchResponse> {
+        val response =
+            khttp.get("https://readfrom.net/build_in_search/?q=$query") // AJAX, MIGHT ADD QUICK SEARCH
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/quicknovel/providers/ReadfromnetProvider.kt]: $mainUrl

File: app/src/main/java/com/lagradost/quicknovel/providers/ReadfromnetProvider.kt [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+import com.lagradost.quicknovel.*
+import org.jsoup.Jsoup
+import java.util.*
+import kotlin.collections.ArrayList
+import kotlin.math.max
+
+
+class ReadfromnetProvider : MainAPI() {
+    override val name = "ReadFrom.Net"
+    override val mainUrl = "https://readfrom.net/"
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/quicknovel/providers/ReadfromnetProvider.kt]: mainUrl should not end with /

**Review by LagradOst — COMMENTED** [review]
Very minor code changes, mainly using $mainUrl and that mainUrl should never end with /. I have no idea if the providers have an svg version of favicon, but if site provides svg versions of the favicon that would be preferable to use,.

**Comment by LagradOst** [comment]
@KillerDogeEmpire

**Comment by KillerDogeEmpire** [comment]
> @KillerDogeEmpire 

Yes?
I am working on it right now

**Comment by LagradOst** [comment]
LGTM, ready to merge?

**Comment by KillerDogeEmpire** [comment]
Ready to merge

**Pull request merged**